src/Features/Core/Portable/EditAndContinue/DebuggingSession.cs

```diff
@@ -49,12 +49,20 @@ internal sealed class DebuggingSession : IDisposable
         /// <remarks>
         /// The baseline of each updated project is linked to its initial baseline that reads from the on-disk metadata and PDB.
         /// Therefore once an initial baseline is created it needs to be kept alive till the end of the debugging session,
-        /// even whne it's replaced in <see cref="_projectEmitBaselines"/> by a newer baseline.
+        /// even when it's replaced in <see cref="_projectEmitBaselines"/> by a newer baseline.
         /// </remarks>
         private readonly Dictionary<ProjectId, EmitBaseline> _projectEmitBaselines = new();
         private readonly List<IDisposable> _initialBaselineModuleReaders = new();
         private readonly object _projectEmitBaselinesGuard = new();
 
+        /// <summary>
+        /// To avoid accessing metadata/symbol readers that have been disposed,
+        /// read lock is acquired before every operation that may access a baseline module/symbol reader 
+        /// and write lock when the baseline readers are being disposed.
+        /// </summary>
+        private readonly ReaderWriterLockSlim _baselineAccessLock = new();
+        private bool _isDisposed;
```

Comment:
❓ Why not use `ReferenceCountedDisposable<T>` to guarantee concurrently-used resources are not disposed while still in use?

Reply:
I don't see how ReferenceCountedDisposable is applicable here.